src/util/result.h

```diff
@@ -9,41 +9,60 @@
 
 #include <variant>
 
-/*
- * 'BResult' is a generic class useful for wrapping a return object
- * (in case of success) or propagating the error cause.
-*/
+namespace util {
+
+struct Error { bilingual_str message; };
+
+//! The util::Result class provides a standard way for functions to return error
+//! messages in addition to optional result values.
+//!
+//! It is intended for high-level functions that need to report error strings to
+//! end users. Lower-level functions that don't need this error-reporting and
+//! only need error-handling should avoid util::Result and instead use standard
+//! classes like std::optional, std::variant, and std::tuple, or custom structs
+//! and enum types to return function results.
+//!
+//! Usage examples can be found in \example ../test/result_tests.cpp, but in
+//! general code returning `util::Result<T>` values is very similar to code
+//! returning `std::optional<T>` values. Existing functions returning
+//! `std::optional<T>` can be updated to return `util::Result<T>` and return
+//! error strings usually just replacing `return std::nullopt;` with `return
+//! util::Error{error_string};`.
 template<class T>
-class BResult {
+class Result {
 private:
     std::variant<bilingual_str, T> m_variant;
 
+    template <typename FT>
```

Comment:
```suggestion
    template <typename OT>
```

nit: Shouldn't this be named `OT`, as it is used in the move-constructor copying the error from `Result<OT>`?

Reply:
I didn't read your reply, but now it is unused and can be removed:


```diff
diff --git a/src/util/result.h b/src/util/result.h
index d038ffe133..cddc8a0819 100644
--- a/src/util/result.h
+++ b/src/util/result.h
@@ -34,8 +34,6 @@ class Result {
 private:
     std::variant<bilingual_str, T> m_variant;
 
-    template <typename FT>
-    friend class Result;
     template <typename FT>
     friend bilingual_str ErrorString(const Result<FT>& result);

Reply:
re: https://github.com/bitcoin/bitcoin/pull/25721#discussion_r935502938

> I didn't read your reply, but now it is unused and can be removed:

Nice catch. It's gone